bip-0346.md

```diff
@@ -0,0 +1,438 @@
+    * 2/1: Inputs
+      * 00: `TXFS_INOUT_SELECTION_NONE`
+      * 01: `TXFS_INOUT_SELECTION_CURRENT`
+      * 11: `TXFS_INOUT_SELECTION_ALL`
+    * 4/3: Outputs
+      * 00: `TXFS_INOUT_SELECTION_NONE`
+      * 01: `TXFS_INOUT_SELECTION_CURRENT`
+      * 11: `TXFS_INOUT_SELECTION_ALL`
+    * 5: `TXFS_INPUTS_PREVOUTS`
+    * 6: `TXFS_INPUTS_PREV_SCRIPTPUBKEYS | TXFS_INPUTS_PREV_VALUES`
+    * 7: `TXFS_CURRENT_INPUT_CONTROL_BLOCK | TXFS_CURRENT_INPUT_SPENTSCRIPT | TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS`
+    * 8: `TXFS_CURRENT_INPUT_IDX`
+
+    Additionally, it includes `TXFS_VERSION | TXFS_LOCKTIME | TXFS_CONTROL | TXFS_CURRENT_INPUT_TAPROOT_ANNEX | TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_OUTPUTS_ALL`
+
+    These 1-byte selections allow the TxFieldSelector to emulate current
+    signature hashing modes and those defined in BIP-118:
+
+| BIP-341/118 sighash type     | 1-byte TxFieldSelector |
+| :--------------------------- | :--------------------- |
+| `ALL`                        | `0b11111111`           |
+| `SINGLE`                     | `0b11110111`           |
+| `NONE`                       | `0b11110011`           |
+| `ALL|ANYONECANPAY`           | `0b11111101`           |
+| `SINGLE|ANYONECANPAY`        | `0b11110101`           |
+| `NONE|ANYONECANPAY`          | `0b11110001`           |
+| `ALL|ANYPREVOUT`             | `0b11101101`           |
+| `SINGLE|ANYPREVOUT`          | `0b11100101`           |
+| `NONE|ANYPREVOUT`            | `0b11100001`           |
+| `ALL|ANYPREVOUTANYSCRIPT`    | `0b11001101`           |
+| `SINGLE|ANYPREVOUTANYSCRIPT` | `0b11000101`           |
+| `NONE|ANYPREVOUTANYSCRIPT`   | `0b11000001`           |
+
+* The first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:
+  * 1: version (`TXFS_VERSION`)
+  * 2: locktime (`TXFS_LOCKTIME`)
+  * 3: current input index (`TXFS_CURRENT_INPUT_IDX`)
+  * 4: current input control block (or empty) (`TXFS_CURRENT_INPUT_CONTROL_BLOCK`)
+  * 5: current input spent script (i.e. witness script or tapscript) (`TXFS_CURRENT_INPUT_SPENTSCRIPT`)
+  * 6: current script last `OP_CODESEPARATOR` position (or 0xffffffff)
+    (`TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS`)
+  * 7: current input annex (or empty) (`TXFS_CURRENT_INPUT_TAPROOT_ANNEX`)
+  * 8: `TXFS_CONTROL` (i.e. include TxFieldSelector into hash)
+
+* The highest bit of the first byte (`TXFS_CONTROL`), we will call the
+  "control bit", and it can be used to control the behavior of the opcode. For
+  `OP_TXHASH` and `OP_CHECKTXHASHVERIFY`, the control bit is used to determine
+  whether the TxFieldSelector itself has to be included in the resulting hash.
+  (For potential other uses of the TxFieldSelector (like a hypothetical
+  `OP_TX`), this bit can be repurposed.)
+
+* The second byte will be used to indicate fields from the inputs and outputs.
+  If there is only a single byte present, no information from the inputs and
```

Comment:
I’m probably missing something, but the option for the TxFieldSelector to only have a single byte here appears to be at odds with the special case above in line 83 "If the TxFieldSelector has exactly 1 byte, […]"